pkg/parse/root.go

```diff
@@ -457,18 +463,30 @@ func setSyncStatusErrors(syncStatus *v1beta1.Status, cse []v1beta1.ConfigSyncErr
 	syncStatus.Sync.Errors = cse[0 : len(cse)/denominator]
 }
 
-// summarizeErrors summarizes the errors from `sourceStatus` and `syncStatus`, and returns an ErrorSource slice and an ErrorSummary.
-func summarizeErrors(sourceStatus v1beta1.SourceStatus, syncStatus v1beta1.SyncStatus) ([]v1beta1.ErrorSource, *v1beta1.ErrorSummary) {
+// summarizeErrorsForCommit summarizes the source, rendering, and sync errors
+// for a specific commit.
+//
+// Since we don't keep errors for old commits after that stage has started
+// working on a new commit, this process is a little bit lossy.
+func summarizeErrorsForCommit(sourceStatus v1beta1.SourceStatus, renderingStatus v1beta1.RenderingStatus, syncStatus v1beta1.SyncStatus, commit string) ([]v1beta1.ErrorSource, *v1beta1.ErrorSummary) {
 	var errorSources []v1beta1.ErrorSource
-	if len(sourceStatus.Errors) > 0 {
+	var summaries []*v1beta1.ErrorSummary
+
+	if sourceStatus.Commit == commit && len(sourceStatus.Errors) > 0 {
```

Comment:
So this means we might have an error for a previous commit on the `renderingStatus`, but the `Syncing` condition will only contain the `errors`/`errorSummary` for the current commit? That makes sense, given that the `commit` is a condition field.

Reply:
That's actually how it is today, because the old code here didn't include any rendering errors and is only called from `setSyncStatusWithRetries`. 

I assume that's because it was originally only sync errors, but then source errors were added when we added non-blocking errors to handle the PoCo case that removes CRs without live CRDs from the apply set.

This new `summarizeErrorsForCommit` is more robust because it includes defensive logic to only ever include errors for the specified commit. But I also moved the commit check from the run loop to `setSyncStatusWithRetries`, which means  `summarizeErrorsForCommit` is actually only ever called when the commits all match. So the change here is redundant/defensive, but I think it also makes the code a little easier to understand and might prevent breakage if we add other updates in the future, like for source or rendering in-progress.

Ideally, the update code would look similar for each stage, but it's all grown very organically and drifted quite a bit just to fix bugs as they are discovered. I'd really love to go in and clean it all up, but it never seems to be top priority.